pkg/internal/tests/cross-tests/impl/hclwrite.go

```diff
@@ -0,0 +1,160 @@
+	lifecycleArgs lifecycleArgs
+}
+
+type WriteResourceOption func(*writeResourceOptions)
+
+func WithCreateBeforeDestroy(createBeforeDestroy bool) WriteResourceOption {
+	return func(o *writeResourceOptions) {
+		o.lifecycleArgs.CreateBeforeDestroy = createBeforeDestroy
+	}
+}
+
+// WriteResource writes a resource declaration to an HCL file.
+//
+// Note that unknowns are not yet supported in cty.Value, it will error out if found.
+func WriteResource(
+	w io.Writer, schema ShimHCLSchema, resourceType, resourceName string, config map[string]cty.Value,
+	opts ...WriteResourceOption,
+) error {
+	if !cty.ObjectVal(config).IsWhollyKnown() {
+		return fmt.Errorf("WriteHCL cannot yet write unknowns")
+	}
+	o := &writeResourceOptions{}
+	for _, opt := range opts {
+		opt(o)
+	}
+
+	if config == nil {
+		config = map[string]cty.Value{}
+	}
+
+	if o.lifecycleArgs.CreateBeforeDestroy {
+		config["lifecycle"] = cty.ObjectVal(map[string]cty.Value{
+			"create_before_destroy": cty.True,
+		})
+	}
+
+	f := hclwrite.NewEmptyFile()
+	block := f.Body().AppendNewBlock("resource", []string{resourceType, resourceName})
+	writeBlock(block.Body(), schema, config)
+	_, err := f.WriteTo(w)
+	return err
+}
+
+type lifecycleBlock struct{}
+
+var _ ShimHCLBlock = &lifecycleBlock{}
+
+func (b *lifecycleBlock) GetNestingMode() Nesting {
+	return NestingSingle
+}
+
+func (b *lifecycleBlock) Attributes() map[string]ShimHCLAttribute {
+	return map[string]ShimHCLAttribute{
+		"create_before_destroy": cty.Bool,
+	}
+}
+
+func (b *lifecycleBlock) Blocks() map[string]ShimHCLBlock {
+	return map[string]ShimHCLBlock{}
+}
+
+func writeBlock(body *hclwrite.Body, schema ShimHCLSchema, config map[string]cty.Value) {
+	attributeList := make([]string, 0, len(schema.Attributes()))
+	for key := range schema.Attributes() {
+		attributeList = append(attributeList, key)
+	}
+	sort.Strings(attributeList)
+	for _, key := range attributeList {
+		v, ok := config[key]
+		if !ok {
+			continue
+		}
+		body.SetAttributeValue(key, v)
+	}
+
+	blockList := make([]string, 0, len(schema.Blocks()))
+	for key := range schema.Blocks() {
+		blockList = append(blockList, key)
+	}
+	sort.Strings(blockList)
+
+	for _, key := range blockList {
+		block := schema.Blocks()[key]
+		if v, ok := config[key]; !ok || v.IsNull() {
+			continue
+		}
+
+		switch block.GetNestingMode() {
+		case NestingSingle:
+			newBlock := body.AppendNewBlock(key, nil)
+			writeBlock(newBlock.Body(), block, config[key].AsValueMap())
+		case NestingList, NestingSet:
+			for _, elem := range config[key].AsValueSlice() {
+				newBlock := body.AppendNewBlock(key, nil)
+				writeBlock(newBlock.Body(), block, elem.AsValueMap())
+			}
+		default:
+			contract.Failf("unexpected nesting mode %v", block.GetNestingMode())
+		}
+	}
+
+	// lifecycle block
+	if _, ok := config["lifecycle"]; ok {
+		newBlock := body.AppendNewBlock("lifecycle", nil)
+		writeBlock(newBlock.Body(), &lifecycleBlock{}, config["lifecycle"].AsValueMap())
+	}
```

Comment:
I'm assuming that we only want `"lifecycle"` to be special at the top level. This will treat it specially at all levels. Given that, can we move this speciality out of `writeBlock` and into `writeResource`. `lifecycle` can just be an unconditional addition to the set of blocks at the top level.